pkg/api/jobs.go

```diff
@@ -168,115 +168,79 @@ func PrintDBJobsReport(w http.ResponseWriter, req *http.Request,
+        current_passes * 100.0 / NULLIF(current_runs, 0) AS current_pass_percentage,
+        (current_passes + current_infra_fails) * 100.0 / NULLIF(current_runs, 0) AS current_projected_pass_percentage,
+        current_fails * 100.0 / NULLIF(current_runs, 0) AS current_failure_percentage,
+        previous_passes * 100.0 / NULLIF(previous_runs, 0) AS previous_pass_percentage,
+        (previous_passes + previous_infra_fails) * 100.0 / NULLIF(previous_runs, 0) AS previous_projected_pass_percentage,
+        previous_failures * 100.0 / NULLIF(previous_runs, 0) AS previous_failure_percentage,
+        (current_passes * 100.0 / NULLIF(current_runs, 0)) - (previous_passes * 100.0 / NULLIF(previous_runs, 0)) AS net_improvement
+FROM results
+JOIN prow_jobs ON prow_jobs.name = results.pj_name
+`
+	r := dbc.DB.Raw(jobsQuery).Scan(&jobReports)
+	if r.Error != nil {
+		klog.Error(r.Error)
+		return []apitype.Job{}, r.Error
+	}
 
-		if jobRun.Succeeded {
-			if jobRun.Timestamp.After(boundaryDate) {
-				jobAcc.CurrentPasses++
-			} else {
-				jobAcc.PreviousPasses++
+	// Apply filtering to what we pulled from the db. Perfect world we'd incorporate this into the query instead.
+	filteredJobReports := make([]apitype.Job, 0, len(jobReports))
+	for _, jobReport := range jobReports {
+		if filter != nil {
+			include, err := filter.Filter(jobReport)
+			if err != nil {
+				return []apitype.Job{}, err
 			}
-		}
 
-		if jobRun.Failed {
-			if jobRun.Timestamp.After(boundaryDate) {
-				jobAcc.CurrentFails++
-				if jobRun.InfrastructureFailure {
-					jobAcc.CurrentInfraFails++
-				}
-			} else {
-				jobAcc.PreviousFails++
-				if jobRun.InfrastructureFailure {
-					jobAcc.PreviousInfraFails++
-				}
+			if !include {
+				continue
 			}
 		}
 
-	}
-
-	// Now that we've processed all job results, calculate percentages on the results we'll return:
-	finalAPIJobResult := make([]apitype.Job, 0, len(apiJobResults))
-	for _, v := range apiJobResults {
-		if v.CurrentRuns > 0 {
-			v.CurrentPassPercentage = float64(v.CurrentPasses) / float64(v.CurrentRuns) * 100
-			v.CurrentProjectedPassPercentage = float64(v.CurrentPasses+v.CurrentInfraFails) / float64(v.CurrentRuns) * 100
-		}
-		if v.PreviousRuns > 0 {
-			v.PreviousPassPercentage = float64(v.PreviousPasses) / float64(v.PreviousRuns) * 100
-			v.PreviousProjectedPassPercentage = float64(v.PreviousPasses+v.PreviousInfraFails) / float64(v.PreviousRuns) * 100
-		}
-		v.NetImprovement = v.CurrentPassPercentage - v.PreviousPassPercentage
-		finalAPIJobResult = append(finalAPIJobResult, *v)
+		filteredJobReports = append(filteredJobReports, jobReport)
 	}
 
 	elapsed := time.Since(now)
-	klog.Infof("BuildJobResult completed in %s with %d results", elapsed, len(finalAPIJobResult))
+	klog.Infof("BuildJobResult completed in %s with %d results from db, filtered down to %s", elapsed, len(jobReports), len(filteredJobReports))
 
-	return finalAPIJobResult, nil
+	return filteredJobReports, nil
```

Comment:
I think you'll want to call `sort(req).limit(req))` on this

Reply:
Looks like I accidentally did that already, this func is called by PrintDBJobsReport which has the request param and does the sort/limit.